File: app/src/main/java/fr/erictruong/material/training/ui/components/lists_controls/IconSwitchListControlActivity.java
package fr.erictruong.material.training.ui.components.lists_controls;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import fr.erictruong.android.core.activity.RecyclerActivity;
import fr.erictruong.android.lists.MaterialListAdapter;
import fr.erictruong.android.lists.OnActionListener;
import fr.erictruong.android.lists.OnCheckActionListener;
import fr.erictruong.android.lists.item.IconSwitchItem;
import fr.erictruong.android.lists.item.MaterialListItem;
import fr.erictruong.material.training.R;
import fr.erictruong.material.training.model.DummyModel;

public class IconSwitchListControlActivity extends RecyclerActivity {

    private static final int ITEM_COUNT = 42;

    private static final DummyModel MODEL = new DummyModel(0, "http://placehold.it/100/888/888/", "Primary text", "Secondary text", "Tertiary text");

    public static void start(Context context) {
        Intent starter = new Intent(context, IconSwitchListControlActivity.class);
        context.startActivity(starter);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final RecyclerView recyclerView = getRecyclerView();

        MaterialListAdapter adapter = new MaterialListAdapter();
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        List<MaterialListItem> items = new ArrayList<>(ITEM_COUNT);
        for (int i = 0; i < ITEM_COUNT; i++) {
            MaterialListItem item = new IconSwitchItem.Builder()
                    .id(i)
                    .icon(R.drawable.gray)
                    .text1(MODEL.getText1())
                    .action(new OnActionListener<IconSwitchItem>() {
                        @Override
                        public void onAction(View v, IconSwitchItem item) {
                            Snackbar.make(getRecyclerView(), "Action #" + item.getId(), Snackbar.LENGTH_SHORT).show();
                        }
                    })
                    .actionSecondary(new OnCheckActionListener<IconSwitchItem>() {
                        @Override
                        public void onAction(View v, IconSwitchItem item) {
                            Snackbar.make(getRecyclerView(), "Secondary action #" + item.getId(), Snackbar.LENGTH_SHORT).show();
                        }
                    })
                    .build();
            items.add(item);
        }
        adapter.swapData(items);
    }
}

File: app/src/main/java/fr/erictruong/material/training/ui/components/lists_controls/IconSwitchListControlActivity.kt
package fr.erictruong.material.training.ui.components.lists_controls

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import fr.erictruong.android.cards.Bindable
import fr.erictruong.android.core.activity.RecyclerActivity
import fr.erictruong.android.lists.holder.IconSwitchViewHolder
import fr.erictruong.android.lists.stub.*
import fr.erictruong.material.training.domain.DummyModel
import java.util.*

class IconSwitchListControlActivity : RecyclerActivity() {

    companion object {
        val KEY_ITEMS = "key_items"
    }

    var items = ArrayList<DummyModel>()
    val adapter = IconSwitchAdapter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adapter

        if (savedInstanceState == null) {
            (0..24).forEach { i -> items.add(DummyModel(i.toLong())) }
            adapter.swapData(items)
        } else {
            items = savedInstanceState.getParcelableArrayList<DummyModel>(KEY_ITEMS)
            adapter.swapData(items)
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState?.putParcelableArrayList(KEY_ITEMS, items)
    }

    class IconSwitchAdapter() : RecyclerView.Adapter<ViewHolder>() {

        var dataset = mutableListOf<DummyModel>()

        fun swapData(items: List<DummyModel>) {
            dataset.addAll(items)
            notifyDataSetChanged()
        }

        override fun getItemCount(): Int = dataset.size

        override fun getItemId(position: Int): Long = dataset[position].id

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder? = ViewHolder(parent, viewType)

        override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(dataset[position])

        override fun onViewRecycled(holder: ViewHolder) = holder.unbind()
    }

    class ViewHolder : IconSwitchViewHolder, Bindable<DummyModel> {

        val actionStub: ActionStub
        val textStub: TextStub
        val iconStub: IconStub
        val switchStub: SwitchStub

        constructor(parent: ViewGroup, viewType: Int) : super(parent, viewType) {
            actionStub = ActionStub(itemView)
            textStub = TextStub(text1)
            iconStub = IconStub(icon)
            switchStub = SwitchStub(toggle)
        }

        override fun bind(item: DummyModel) {
            actionStub.bind { v -> Snackbar.make(v, "Action - id=${item.id}", Snackbar.LENGTH_SHORT).show() }
            switchStub.bind(item.isSelected) { buttonView, isChecked ->
                item.isSelected = isChecked
                Snackbar.make(buttonView, "Secondary action - id=${item.id}", Snackbar.LENGTH_SHORT).show()
            }
            // TODO: Bind holder to item.
        }

        override fun unbind() {
            actionStub.unbind()
            switchStub.unbind()
            // TODO: Unbind holder.
        }
    }
}
